tooling/ipns/record.go

```diff
@@ -41,10 +51,46 @@ func (i *IpnsRecord) Validity() time.Time {
 }
 
 func (i *IpnsRecord) Valid() error {
-	id, err := peer.Decode(i.key)
+	return ipns.ValidateWithPeerID(i.id, i.pb)
+}
+
+func (i *IpnsRecord) idV1(codec multicodec.Code, base mbase.Encoding) (string, error) {
+	c := peer.ToCid(i.id)
+	c = cid.NewCidV1(uint64(codec), c.Hash())
+	s, err := c.StringOfBase(base)
+	if err != nil {
+		return "", err
+	}
+	return s, nil
+}
+
+func (i *IpnsRecord) IntoCID(codec multicodec.Code, base mbase.Encoding) string {
+	s, err := i.idV1(codec, base)
 	if err != nil {
-		return err
+		panic(err)
+	}
+	return s
+}
+
+func (i *IpnsRecord) IdV0() string {
+	if strings.HasPrefix(i.key, "Qm") || strings.HasPrefix(i.key, "1") {
+		return i.key
 	}
+	panic("not a v0 id")
+	// This code won't work if key starts with "1", this is an identity multihash, which
+	// makes `NewCidV0` panic.
+	// c := peer.ToCid(i.id)
+	// s, err := cid.NewCidV0(c.Hash()).StringOfBase(mbase.Base58BTC)
+	// if err != nil {
+	// 	panic(err)
+	// }
+	// return s
+}
+
+func (i *IpnsRecord) IdV1() string {
+	return i.IntoCID(cid.Libp2pKey, mbase.Base36)
+}
 
-	return ipns.ValidateWithPeerID(id, i.pb)
+func (i *IpnsRecord) B58MH() string {
+	return i.id.String()
```

Comment:
Highlight: is this the correct implementation?

Reply:
cc @hacdias in context of https://github.com/ipfs/boxo/pull/339

Reply:
Thanks for the note, I logged that as an item in #8, we'll review once the PR is merged!